contracts/WooPPV2.sol

```diff
@@ -0,0 +1,371 @@
+        emit FeeManagerUpdated(_feeManager);
+    }
+
+    /* ----- Admin Functions ----- */
+
+    function claimFee() external onlyAdmin {
+        TransferHelper.safeTransfer(quoteToken, address(feeManager), unclaimedFee);
+        unclaimedFee = 0;
+    }
+
+    function setFeeRate(address token, uint16 rate) external onlyAdmin {
+        require(rate <= 1e5, '!rate');
+        tokenInfos[token].feeRate = rate;
+    }
+
+    function pause() external onlyAdmin {
+        super._pause();
+    }
+
+    function unpause() external onlyAdmin {
+        super._unpause();
+    }
+
+    function setAdmin(address addr, bool flag) external onlyAdmin {
+        require(addr != address(0), 'WooPPV2: !admin');
+        isAdmin[addr] = flag;
+        emit AdminUpdated(addr, flag);
+    }
+
+    function deposit(address token, uint256 amount) external onlyAdmin {
+        uint256 balanceBefore = balance(token);
+        TransferHelper.safeTransferFrom(token, msg.sender, address(this), amount);
+        uint256 amountReceived = balance(token).sub(balanceBefore);
+        require(amountReceived >= amount, 'AMOUNT_INSUFF');
+
+        tokenInfos[token].reserve = uint192(tokenInfos[token].reserve.add(amount));
+
+        emit Deposit(token, msg.sender, amount);
+    }
+
+    function withdraw(address token, uint256 amount) public onlyAdmin {
+        require(tokenInfos[token].reserve >= amount, 'WooPPV2: !amount');
+        TransferHelper.safeTransfer(token, _OWNER_, amount);
+        tokenInfos[token].reserve = uint192(tokenInfos[token].reserve.sub(amount));
+        emit Withdraw(token, _OWNER_, amount);
+    }
+
+    function withdrawAll(address token) external onlyAdmin {
+        withdraw(token, poolSize(token));
+    }
+
+    /* ----- Private Functions ----- */
+
+    function _updateReserve(address baseToken) private {
+        require(
+            balance(baseToken) > tokenInfos[baseToken].reserve || balance(quoteToken) > tokenInfos[quoteToken].reserve,
+            'WooPPV2: !BALANCE'
+        );
+
+        // TODO: how to handle the accidental transferred funds?
+        tokenInfos[baseToken].reserve = uint192(balance(baseToken));
+        tokenInfos[quoteToken].reserve = uint192(balance(quoteToken).sub(unclaimedFee));
+    }
+
+    function getQuoteAmountSellBase(address baseToken, uint256 baseAmount)
+        private
+        view
+        returns (uint256 quoteAmount, uint256 newPrice)
+    {
+        IWooracleV2.State memory state = IWooracleV2(wooracle).state(baseToken);
+        require(state.woFeasible, 'WooPPV2: ORACLE_PRICE_NOT_FEASIBLE');
+
+        Decimals memory decs = _decimals(baseToken);
+
+        // quoteAmount = baseAmount * oracle.price * (1 - oracle.k * baseAmount * oracle.price - oracle.spread)
+        {
+            uint256 coef = uint256(1e18) -
+                ((uint256(state.coeff) * baseAmount * state.price) / decs.baseDec / decs.priceDec) -
+                state.spread;
```

Comment:
why not use `div()`, `mul()`, `sub()`?

Reply:
will migrate to 0.8.  Using those methods will cause stack too deep.